trl/trainer/srpo_trainer.py

```diff
@@ -0,0 +1,1525 @@
+        improve_to_rejected_given_rejected = policy_logps["improve_to_rejected_given_rejected"].to(
+            self.accelerator.device
+        )
+        improve_to_chosen_given_chosen = policy_logps["improve_to_chosen_given_chosen"].to(self.accelerator.device)
+        improve_to_rejected_given_chosen = policy_logps["improve_to_rejected_given_chosen"].to(self.accelerator.device)
+        chosen_logps = policy_logps["chosen"].to(self.accelerator.device)
+        rejected_logps = policy_logps["rejected"].to(self.accelerator.device)
+
+        ref_improve_to_chosen_given_rejected = reference_logps["improve_to_chosen_given_rejected"].to(
+            self.accelerator.device
+        )
+        ref_improve_to_rejected_given_rejected = reference_logps["improve_to_rejected_given_rejected"].to(
+            self.accelerator.device
+        )
+        ref_improve_to_chosen_given_chosen = reference_logps["improve_to_chosen_given_chosen"].to(
+            self.accelerator.device
+        )
+        ref_improve_to_rejected_given_chosen = reference_logps["improve_to_rejected_given_chosen"].to(
+            self.accelerator.device
+        )
+        ref_chosen_logps = reference_logps["chosen"].to(self.accelerator.device)
+        ref_rejected_logps = reference_logps["rejected"].to(self.accelerator.device)
+
+        rejected_to_chosen_improvement_ratio = improve_to_chosen_given_rejected - ref_improve_to_chosen_given_rejected
+        rejected_to_rejected_improvement_ratio = (
+            improve_to_rejected_given_rejected - ref_improve_to_rejected_given_rejected
+        )
+        given_rejected = (
+            0.5 - self.beta * (rejected_to_chosen_improvement_ratio - rejected_to_rejected_improvement_ratio)
+        ) ** 2
+
+        chosen_to_chosen_improvement_ratio = improve_to_chosen_given_chosen - ref_improve_to_chosen_given_chosen
+
+        chosen_to_rejected_improvement_ratio = improve_to_rejected_given_chosen - ref_improve_to_rejected_given_chosen
+        given_chosen = (
+            0.5 - self.beta * (chosen_to_chosen_improvement_ratio - chosen_to_rejected_improvement_ratio)
+        ) ** 2
+        self_improvement_loss = given_rejected + given_chosen
+
+        chosen_zero_ratio = chosen_logps - ref_chosen_logps
+        rejected_zero_ratio = rejected_logps - ref_rejected_logps
+        log_terms = (
+            rejected_to_chosen_improvement_ratio
+            + chosen_zero_ratio
+            - (chosen_to_rejected_improvement_ratio + rejected_zero_ratio)
+        )
+        generative_loss = (self.beta * log_terms - 1) ** 2
+        losses = (1 - self.alpha) * generative_loss + self.alpha * self_improvement_loss
+        # TODO Deal with reference_free
+        # if self.reference_free:
+        #     ref_logratios = torch.tensor([0], dtype=pi_logratios.dtype, device=pi_logratios.device)
+        # else:
+        #     ref_logratios = reference_chosen_logps - reference_rejected_logps
```

Comment:
Interesting how clearly this reduces to DPO + additional metadata. Nice to see it in here.